airbyte-cdk/python/airbyte_cdk/sources/file_based/stream/default_file_based_stream.py

```diff
@@ -178,7 +178,12 @@ def _get_raw_json_schema(self) -> JsonSchema:
             if total_n_files == 0:
                 raise NoFilesMatchingError(FileBasedSourceError.EMPTY_STREAM, stream=self.name)
 
-            max_n_files_for_schema_inference = self._discovery_policy.max_n_files_for_schema_inference
+            max_n_files_for_schema_inference = min(
```

Comment:
To continue to allow `DefaultDiscoveryPolicy` to be the point of truth for the max # of files used, WDYT about turning `DefaultDiscoveryPolicy.max_n_files_for_schema_inference` into a method instead of a property, and then passing the stream in? `DefaultDiscoveryPolicy.max_n_files_for_schema_inference` can then look for overrides and calculate the max number of files to use as `min(default maximum, maximum as specified by the parser)` to hide that logic away from the stream itself.

(I'm suggesting passing in the stream instead of the parser itself in case there are other non-parser considerations we want to take into account in the future.)

I'd also rename the parser's  `override_max_n_files_for_schema_inference` to `parser_max_n_files_for_schema_inference`.